alf/algorithms/hypernetwork_algorithm.py

```diff
@@ -0,0 +1,430 @@
+            input_tensor_spec=input_tensor_spec,
+            conv_layer_params=conv_layer_params,
+            fc_layer_params=fc_layer_params,
+            activation=activation,
+            last_layer_param=last_layer_param,
+            last_activation=last_activation)
+
+        gen_output_dim = param_net.param_length
+        noise_spec = TensorSpec(shape=(noise_dim, ))
+        net = EncodingNetwork(
+            noise_spec,
+            fc_layer_params=hidden_layers,
+            use_fc_bn=use_fc_bn,
+            last_layer_size=gen_output_dim,
+            last_activation=math_ops.identity,
+            name="Generator")
+
+        if logging_network:
+            logging.info("Generated network")
+            logging.info("-" * 68)
+            logging.info(param_net)
+
+            logging.info("Generator network")
+            logging.info("-" * 68)
+            logging.info(net)
+
+        if par_vi == 'svgd':
+            par_vi = 'svgd3'
+
+        self._generator = Generator(
+            gen_output_dim,
+            noise_dim=noise_dim,
+            net=net,
+            entropy_regularization=entropy_regularization,
+            par_vi=par_vi,
+            optimizer=optimizer,
+            name=name)
+
+        self._param_net = param_net
+        self._num_particles = num_particles
+        self._entropy_regularization = entropy_regularization
+        self._train_loader = None
+        self._test_loader = None
+        self._use_fc_bn = use_fc_bn
+        self._loss_type = loss_type
+        self._logging_training = logging_training
+        self._logging_evaluate = logging_evaluate
+        self._config = config
+        assert (voting in ['soft', 'hard'
+                           ]), ("voting only supports \"soft\" and \"hard\"")
+        self._voting = voting
+        if loss_type == 'classification':
+            self._vote = self._classification_vote
+        elif loss_type == 'regression':
+            self._vote = self._regression_vote
+        else:
+            raise ValueError("Unsupported loss_type: %s" % loss_type)
+
+    def set_data_loader(self, train_loader, test_loader=None):
+        """Set data loadder for training and testing."""
+        self._train_loader = train_loader
+        self._test_loader = test_loader
+        self._entropy_regularization = 1 / len(train_loader)
+
+    def set_num_particles(self, num_particles):
+        """Set the number of particles to sample through one forward
+        pass of the hypernetwork. """
+        self._num_particles = num_particles
+
+    @property
+    def num_particles(self):
+        """number of sampled particles. """
+        return self._num_particles
+
+    def sample_parameters(self, noise=None, num_particles=None, training=True):
+        """Sample parameters for an ensemble of networks. 
+        
+        Args:
+            noise (Tensor): input noise to self._generator.
+            num_particles (int): number of sampled particles.
```

Comment:
comment num_particles==None, here and at other places

Reply:
updated.